Add support for feature unavailable error

RonEld opened the pull request:
## Description
Add `MBEDTLS_ERR_XXX_FEATURE_UNAVAILABLE` to modules that:
* support alternative implementation
* don't alreaedy have the error
* have features that can be not supported by HW acceleration

Skip this error in:
* tests
* examples

resolves #1447 among other features

## Status
**IN DEVELOPMENT**

## Requires Backporting
NO  

## Additional comments
Any additional information that could be of interest

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

RonEld commented:
@Patater Thank you for your comments.
I addressed your comments, but I think that at this stage, we should discuss mostly on the header files, and what actual functions should return the not supported errors.

I added support for this error in the test suites, except in x509, where I tihnk that if and ECP curve is not supported( even if user didn't disable the curve in configuration), or the RSA key size is not supported by the HW, then x509 verification should fail. I also think that x509 support for the unsupported feature, requires further investigation and design, that extents the scope of this PR.

Note: test suites realted to ctr_drbg ( `test_suite_ctr_drbg` and some tests in `test_suite_rsa` fail when AES returns `MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE`, because ctr_drbg mandates AES256. This is addressed in #1183 )

RonEld commented:
As for tests, we should definitely have some configurations with alternative implementations. 
At the moment, I manually changed return statements of functions I thought should return the error code, for testing my implementation.

RonEld commented:
I have update the sample applications which I think should ignore the FEAUTRE_UNAVAILABLE error and continue their operation.

Patater commented on programs/test/benchmark.c:
```diff
@@ -418,10 +444,16 @@ int main( int argc, char *argv[] )
             memset( tmp, 0, sizeof( tmp ) );
-            mbedtls_aes_setkey_enc( &aes, tmp, keysize );
+            ret = mbedtls_aes_setkey_enc( &aes, tmp, keysize );
+            if( ret  != 0 )
```
Remove extra space after `ret`.

Here and in other places.

Patater commented on programs/test/benchmark.c:
```diff
@@ -589,19 +651,33 @@ int main( int argc, char *argv[] )
         mbedtls_ctr_drbg_context ctr_drbg;
 
         mbedtls_ctr_drbg_init( &ctr_drbg );
+        mbedtls_snprintf( title, sizeof( title ), "CTR_DRBG (NOPR)" );
```
Why change to `mbedtls_snprintf` instead of passing the string directly to `TIME_AND_TSC`?

RonEld replied:
so in case of error, we will print the title before we call `PRINT_ERROR`.
This will change, after I rebase, and the code from https://github.com/ARMmbed/mbedtls/commit/51d7cfe026fa3f33d86a80d27ea9013830f33992 will be introduced

Patater commented on programs/test/benchmark.c:
```diff
@@ -589,19 +651,33 @@ int main( int argc, char *argv[] )
+        else
+        {
+            TIME_AND_TSC( title,
+                    ret = mbedtls_ctr_drbg_random( &ctr_drbg, buf, BUFSIZE )  );
```
Please revise wrapping.

Patater commented on programs/test/benchmark.c:
```diff
@@ -618,41 +694,69 @@ int main( int argc, char *argv[] )
+        else
+        {
+            TIME_AND_TSC( title,
+                    ret = mbedtls_hmac_drbg_random( &hmac_drbg, buf, BUFSIZE ) );
```
Please revise wrapping.

Here and in other similar places.

Patater commented on programs/test/benchmark.c:
```diff
@@ -589,19 +651,33 @@ int main( int argc, char *argv[] )
+        if( ret  != 0 )
+        {
+            mbedtls_printf( HEADER_FORMAT, title );
+            PRINT_ERROR;
```
Why don't we call `mbedtls_exit(1)` anymore?

Here and in other places.

Patater commented on programs/test/benchmark.c:
```diff
@@ -618,41 +694,69 @@ int main( int argc, char *argv[] )
+        else
+        {
+            mbedtls_hmac_drbg_set_prediction_resistance( &hmac_drbg,
+                                                 MBEDTLS_HMAC_DRBG_PR_ON );
```
Please revise wrapping.

Patater commented on programs/test/benchmark.c:
```diff
@@ -668,7 +772,13 @@ int main( int argc, char *argv[] )
             mbedtls_snprintf( title, sizeof( title ), "RSA-%d", keysize );
 
             mbedtls_rsa_init( &rsa, MBEDTLS_RSA_PKCS_V15, 0 );
```
Can `mbedtls_rsa_init` fail with `FEATURE_UNSUPPORTED` or any other errors?

RonEld replied:
perhaps, but `rsa_init()` returns void, and according to the coding standards:
>The `init()` function must always return `void`

Patater replied:
OK

Patater commented on programs/test/benchmark.c:
```diff
@@ -752,13 +868,18 @@ int main( int argc, char *argv[] )
+            {
+                mbedtls_printf( HEADER_FORMAT, title );
+                PRINT_ERROR;
+                continue;
```
Should we call `mbedtls_exit( 1 )` here as we did before?

The same goes for other similar places.

RonEld replied:
I prefer not to exit in case of error. Since the code is quite generic, and we don't have a common error code for all feature unavailable, we can't distinguish between a feature unavailable error and a real error, so I would prefer just printing it and go to next step

Patater replied:
OK, that's a change in behavior. We should do that in its own commit so we can review it and potentially merge it independently from your other changes. Then you can build on top of that change in further commits. I think Andres unified return codes in another PR if not for tests, at least Mbed OS examples; we could do something similar here.

RonEld replied:
ok, I'll return the exit, but this means that when a FEATURE_UNAVAILABLE error returns, we also exit

RonEld replied:
Although it's a change in behaviour, the whole idea of this PR is to support the FEATURE_UNAVAILABLE error, in the applications themselves. If we exit when we get this valid error, it loses the idea of this PR

Patater commented on tests/suites/test_suite_aes.function:
```diff
@@ -61,10 +68,16 @@ void aes_decrypt_ecb( char *hex_key_string, char *hex_src_string,
+    ret = mbedtls_aes_setkey_dec( &ctx, key_str, key_len * 8 );
+    if( ret == MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE )
+        goto exit;
+    TEST_ASSERT(  ret == setkey_result );
```
Remove extra space before `ret`.

Patater commented on tests/suites/test_suite_aes.function:
```diff
@@ -61,10 +68,16 @@ void aes_decrypt_ecb( char *hex_key_string, char *hex_src_string,
     if( setkey_result == 0 )
     {
-        TEST_ASSERT( mbedtls_aes_crypt_ecb( &ctx, MBEDTLS_AES_DECRYPT, src_str, output ) == 0 );
+        ret =  mbedtls_aes_crypt_ecb( &ctx, MBEDTLS_AES_DECRYPT, src_str, output );
```
Remove extra space after `=`.

Patater commented on tests/suites/test_suite_aes.function:
```diff
@@ -99,8 +112,14 @@ void aes_encrypt_cbc( char *hex_key_string, char *hex_iv_string,
+    ret = mbedtls_aes_crypt_cbc( &ctx, MBEDTLS_AES_ENCRYPT, data_len, iv_str, src_str, output );
+    if( ret == MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE )
+        goto exit;
+    TEST_ASSERT(  ret == cbc_result );
```
Remove extra space before `ret`.

Patater commented on tests/suites/test_suite_cipher.function:
```diff
@@ -163,8 +175,20 @@ void enc_dec_buf( int cipher_id, char *cipher_string, int key_len,
+        ( ret == MBEDTLS_ERR_CAMELLIA_FEATURE_UNAVAILABLE ) ||
+        ( ret == MBEDTLS_ERR_GCM_FEATURE_UNAVAILABLE ) ||
+        ( ret == MBEDTLS_ERR_DES_FEATURE_UNAVAILABLE ) )
+        goto exit;
```
No action required, but just curious:
Is there a way to say a test was skipped yet?

RonEld replied:
not with this infrastructure, unfortunately.
The skip is done in the `main_test.function`, and until we have a common `FEATURE_UNAVAILABLE` error, we can't determine this.

RonEld commented on include/mbedtls/aes.h:
```diff
@@ -122,6 +122,8 @@ void mbedtls_aes_free( mbedtls_aes_context *ctx );
  * \return         \c 0 on success.
  * \return         #MBEDTLS_ERR_AES_INVALID_KEY_LENGTH on failure.
+ * \return         #MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE if the
+ *                 alternative implementation doesn't support the key size.
```
What other reason for a HW acceleration not to support a feature when calling `mbedtls_aes_setkey_enc()`? 
The same goes for other operations. I tried to be as specific as possible, so we know what to test and what not

Patater replied:
We as a library can't know the many ways in which hardware could choose not to support a certain feature. We are bound to get things wrong for some hardware. I'd rather us try to avoid specifics as much as possible, to give hardware freedom to err however they see fit.

RonEld replied:
But we can define what is mandatory and what not. A developer that wants to develop and alternative implementation will need some guidance on what to return when.

RonEld replied:
OK

RonEld commented on include/mbedtls/ccm.h:
```diff
@@ -85,6 +85,8 @@ void mbedtls_ccm_init( mbedtls_ccm_context *ctx );
  * \return          \c 0 on success.
  * \return          A CCM or cipher-specific error code on failure.
+ * \return          #MBEDTLS_ERR_CCM_FEATURE_UNAVAILABLE if the
+ *                  alternative implementation doesn't support the key size or cipher mode.
```
For what reason? The features that an alternative implementation cannot support are the key size and the cipher type.
Both are set in `mbedtls_ccm_setkey()`.
Perhaps the length is too big for the HW to handle.

Patater replied:
There could be a glitch in the power supply, or the hardware is in use doing another operation and can't begin a new operation, or hardware has a bug and the driver for the software is preventing use of the hardware for this function as a "solution" to prevent people from using buggy hardware.

RonEld replied:
we have [`MBEDTLS_ERR_AES_HW_ACCEL_FAILED`](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/aes.h#L64) for any HW failure, which is a real failure. The FEATURE_UNAVAILBE error is a valid error, that tells the user that this feature is not suipported by the HW.

RonEld commented on library/aes.c:
```diff
@@ -1336,12 +1336,7 @@ int mbedtls_aes_self_test( int verbose )
-         * MBEDTLS_AES_ALT is defined.
-         */
-        if( ret == MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE && keybits == 192 )
+        if( ret == MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE )
```
If a selftest should be handled in the alternative implementation as well, most of the time it will end up with copy \ pasting code.
Perhaps we should have some hooks for selftest callbacks, as suggested in https://github.com/ARMmbed/mbedtls/pull/1598#discussion_r187421900 but I am not sure it is part of the scope of this PR

RonEld replied:
Perhaps we should just remove the changes in the self test in this PR, since they will be removed when we add support for  an alternative self test?

Patater replied:
Yes, I'm recommending we at least "remove the changes in the self test in this PR, since they will be removed when we add support for an alternative self test". Additionally, it shouldn't be too hard to make a new commit to make self-tests ALT-implementable in this PR. We can add hooks later so folks can avoid copy paste, but that needs further design work and is out of scope for this PR; a simple change would allow the self-tests to be ALT-implementable.

RonEld replied:
>it shouldn't be too hard to make a new commit to make self-tests ALT-implementable in this PR

Right, it's just maintenance, having an additional self-test_ALT for every module. Having one self-test_ALT is also not recommended, since most of the time, only a few self tests will actually require alternative implementation

RonEld replied:
I agree that self tests should pass, but different hardware accept different inputs, for example different buffer alignments.
I discussed with @Patater offline, and we will keep the current self test changes, accepting the `FEATURE_UNAVALILABLE`. (see [comment](https://github.com/ARMmbed/mbedtls/pull/1716#issuecomment-408849381))

RonEld commented on library/pkparse.c:
```diff
@@ -1297,7 +1297,9 @@ int mbedtls_pk_parse_key( mbedtls_pk_context *pk,
-    if( ret == MBEDTLS_ERR_PK_PASSWORD_MISMATCH )
+    if( ( ret == MBEDTLS_ERR_PK_PASSWORD_MISMATCH ) ||
+        ( ret == MBEDTLS_ERR_PKCS5_FEATURE_UNAVAILABLE ) ||
+        ( ret == MBEDTLS_ERR_DES_FEATURE_UNAVAILABLE ) )
```
When the key file is encrypted with DES, and the specific feature in DES is not supported by the acceleration

RonEld replied:
In cases where it's not PKCS5.
For example:
~~~
PKCS#8 encrypted v2 PBKDF2 DES hmacWithSHA256
~~~

gilles-peskine-arm replied:
Could also be AES or any other cipher, depending on the input data.

RonEld replied:
I only added the errors that rose when I set different `FEATURE_UNAVAILABLE` in different modules

RonEld replied:
As mentioned before, since we only support DES in our encrypted private key files ( and RC ), I don't the overhead of another error, which we don't test

RonEld commented:
@Patater Thank you for your review.
I believe I have addressed all your comments that don't require code change.
To answer your question
>What testing has been done?

I have manually changed the function return codes, to return the relevant `FEATURE_UNAVAILABLE` error, and modified the tests \ programs accordingly.

General notes:
* I agree we should have a test infrastructure to test alternative implementations in our infrastructure, but I am not sure it's part of the scope of this PR.
* If we had a common `FEATURE_UNAVAIlABLE` error(and perhaps other error codes, not relevant for this PR), then the code would be much cleaner, clearer, and better tests.
* I don't think we should be too vague on what conditions should an alternative implementation return the `FEATURE_UNAVAILABLE` error, as it will be more difficult for a test plan. I think that if we understand clearly what can be unsupported by the HW, and what is considered to be the core of the module, and without it, there is no point of defining the module in the configuration, it would be easier for a developer to implement the alternative implementation, and we have a full and clear test \ coverage.

RonEld commented:
@Patater I will rebase, in order to resolve conflicts in the benchmark application, and to let https://github.com/ARMmbed/mbedtls/commit/51d7cfe026fa3f33d86a80d27ea9013830f33992 introduced.

RonEld commented:
Rebased to resolve conflicts in the benchmark application.
Modified benchmark application according to comments.
Changed error codes after error value conflicts surfaced from Rebase.

RonEld commented:
@Patater As discussed offline, we will keep the changes in the self tests.
I have addressed all your other comments.

Patater commented:
Please look into the CI failures.

RonEld commented:
@Patater I fixed the CI failures

RonEld commented on library/aria.c:
```diff
@@ -909,17 +909,36 @@ int mbedtls_aria_self_test( int verbose )
-        mbedtls_aria_crypt_ecb( &ctx, aria_test1_ecb_ct[i], blk );
-        if( memcmp( blk, aria_test1_ecb_pt, MBEDTLS_ARIA_BLOCKSIZE ) != 0 )
+        ret = mbedtls_aria_setkey_dec( &ctx, aria_test1_ecb_key, 128 + 64 * i );
+        if( ret != 0 )
```
I agree, but this means changing the self-tests logic, as the decryption is dependent on the output of the encryption.

RonEld commented on library/pkcs12.c:
```diff
@@ -216,7 +216,10 @@ int mbedtls_pkcs12_pbe( mbedtls_asn1_buf *pbe_params, int mode,
     if( ( ret = mbedtls_cipher_finish( &cipher_ctx, output + olen, &olen ) ) != 0 )
-        ret = MBEDTLS_ERR_PKCS12_PASSWORD_MISMATCH;
+    {
+        if( ret != MBEDTLS_ERR_DES_FEATURE_UNAVAILABLE )
```
There is no PKCS12 error, because there is no alternative for PKCS12.
At the moment, the only PKCS12 PBE types we support are `pbeWithSHAAnd3-KeyTripleDES-CBC` and `pbeWithSHAAnd2-KeyTripleDES-CBC`, this is why the only acceptable FEATURE UNAVAILABLE I see is the DES one.

gilles-peskine-arm replied:
What happens if the user tries to read some data which happens to be encrypted with AES?

RonEld commented on library/pkcs12.c:
```diff
@@ -216,7 +216,10 @@ int mbedtls_pkcs12_pbe( mbedtls_asn1_buf *pbe_params, int mode,
     if( ( ret = mbedtls_cipher_finish( &cipher_ctx, output + olen, &olen ) ) != 0 )
-        ret = MBEDTLS_ERR_PKCS12_PASSWORD_MISMATCH;
+    {
+        if( ret != MBEDTLS_ERR_DES_FEATURE_UNAVAILABLE )
```
I can add these, but since we don't support other ciphers, and don't test it, I don't think we should add the other errors.

gilles-peskine-arm replied:
What do you mean exactly by “we don't support other ciphers”? We don't control the input to the functions and I don't see any code that rejects other ciphers.

RonEld replied:
When we parse the input, we will fail in https://github.com/ARMmbed/mbedtls/blob/development/library/pkparse.c#L1078

gilles-peskine-arm replied:
Ok, but that's fragile: we could add OIDs at any time and that shouldn't suddenly make some code return weird errors.

RonEld replied:
I agree, and this is why I strongly think we should have one common error for feature not supported by alternative implementation( I think we should distinguish between features that Mbed TLS doesn't support and features that Mbed TLS supports but the alternative implemetnation doesn't)

RonEld commented on library/pkcs5.c:
```diff
@@ -222,7 +223,17 @@ int mbedtls_pkcs5_pbes2( const mbedtls_asn1_buf *pbe_params, int mode,
-        ret = MBEDTLS_ERR_PKCS5_PASSWORD_MISMATCH;
+    {
+        /*
+         * PKCS5 supports DES or RC2. RC2 is not supported by Mbed TLS.
```
Since PKCS5 now supports AES(but we still don't), I will add AES.
I don't think that adding other errors that we don't test is good. The code is already becoming difficult to read, since we don't have one common `FEATURE_UNAVAILABLE` error

gilles-peskine-arm replied:
Are we still ruling out adding a `common_errors.h`, by the way? It would make so much sense, the only problem being that we can't replace existing errors so our choice is between piling on bad error codes after bad or having inconsistently sometimes-unified-sometimes-not error codes.

RonEld replied:
I would really want to have one generic error for feature unavailable, It would make the code much cleaner.
However, this is still not approved, and not discussed yet

gilles-peskine-arm commented on library/cmac.c:
```diff
@@ -798,17 +802,30 @@ static int cmac_test_subkeys( int verbose,
-            if( verbose != 0 )
-                mbedtls_printf( "test execution failed\n" );
+            if ( ( ret == MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE ) ||
+                 ( ret == MBEDTLS_ERR_DES_FEATURE_UNAVAILABLE ) )
```
Yes, I meant `CMAC_FEATURE_UNAVAILABLE`
